image/image_linux.go

```diff
@@ -272,302 +273,228 @@ func makeLabel(now time.Time) string {
+	hasModes    bool
+	rootDir     string
+	bootRootDir string
+	seedDir     string
+	label       string
 
-var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) error {
-	if model.Classic() != opts.Classic {
-		return fmt.Errorf("internal error: classic model but classic mode not set")
-	}
-
-	hasModes := model.Grade() != asserts.ModelGradeUnset
-	var rootDir string
-	var bootRootDir string
-	var seedDir string
-	var label string
-	if !hasModes {
-		if opts.Classic {
-			// Classic, PrepareDir is the root dir itself
-			rootDir = opts.PrepareDir
-		} else {
-			// Core 16/18,  writing for the writeable partition
-			rootDir = filepath.Join(opts.PrepareDir, "image")
-			bootRootDir = rootDir
-		}
-		seedDir = dirs.SnapSeedDirUnder(rootDir)
-
-		// validity check target
-		if osutil.FileExists(dirs.SnapStateFileUnder(rootDir)) {
-			return fmt.Errorf("cannot prepare seed over existing system or an already booted image, detected state file %s", dirs.SnapStateFileUnder(rootDir))
-		}
-		if snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDirUnder(rootDir), "*.snap")); len(snaps) > 0 {
-			return fmt.Errorf("expected empty snap dir in rootdir, got: %v", snaps)
-		}
+	w     *seedwriter.Writer
+	model *asserts.Model
+	tsto  *tooling.ToolingStore
+}
 
+func (is *imageSeeder) modelessDirs(opts *Options) error {
+	if opts.Classic {
+		// Classic, PrepareDir is the root dir itself
+		is.rootDir = opts.PrepareDir
 	} else {
-		// Core 20, writing for the system-seed partition
-		seedDir = filepath.Join(opts.PrepareDir, "system-seed")
-		label = makeLabel(time.Now())
-		bootRootDir = seedDir
-
-		// validity check target
-		if systems, _ := filepath.Glob(filepath.Join(seedDir, "systems", "*")); len(systems) > 0 {
-			return fmt.Errorf("expected empty systems dir in system-seed, got: %v", systems)
-		}
+		// Core 16/18,  writing for the writeable partition
+		is.rootDir = filepath.Join(opts.PrepareDir, "image")
+		is.bootRootDir = is.rootDir
 	}
+	is.seedDir = dirs.SnapSeedDirUnder(is.rootDir)
 
-	// TODO: developer database in home or use snapd (but need
-	// a bit more API there, potential issues when crossing stores/series)
-	db, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
-		Backstore: asserts.NewMemoryBackstore(),
-		Trusted:   trusted,
-	})
-	if err != nil {
-		return err
+	// validity check target
+	if osutil.FileExists(dirs.SnapStateFileUnder(is.rootDir)) {
+		return fmt.Errorf("cannot prepare seed over existing system or an already booted image, detected state file %s", dirs.SnapStateFileUnder(is.rootDir))
+	}
+	if snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDirUnder(is.rootDir), "*.snap")); len(snaps) > 0 {
+		return fmt.Errorf("expected empty snap dir in rootdir, got: %v", snaps)
 	}
+	return nil
+}
 
-	wOpts := &seedwriter.Options{
-		SeedDir:        seedDir,
-		Label:          label,
-		DefaultChannel: opts.Channel,
+func (is *imageSeeder) modeDirs(opts *Options) error {
```

Comment:
same issue, also this probably initModesDirs or setModesDirs

Reply:
Same, but `setModesDirs`